pkg/ibmcsidriver/controller.go

```diff
@@ -418,30 +434,161 @@ func (csiCS *CSIControllerServer) ControllerGetCapabilities(ctx context.Context,
+	snapshotTags := map[string]string{
+		"name": snapshotName,
+	}
+	snapshotParameters.SnapshotTags = snapshotTags
+
+	snapshot, err = session.CreateSnapshot(sourceVolumeID, snapshotParameters)
+
+	if err != nil {
+		return nil, commonError.GetCSIError(ctxLogger, commonError.InternalError, requestID, err, "creation")
+	}
+	return createCSISnapshotResponse(*snapshot), nil
 }
 
 // DeleteSnapshot ...
 func (csiCS *CSIControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequest) (*csi.DeleteSnapshotResponse, error) {
 	ctxLogger, requestID := utils.GetContextLogger(ctx, false)
 	// populate requestID in the context
-	_ = context.WithValue(ctx, provider.RequestID, requestID)
+	ctx = context.WithValue(ctx, provider.RequestID, requestID)
+	defer metrics.UpdateDurationFromStart(ctxLogger, "DeleteSnapshot", time.Now())
+	ctxLogger.Info("CSIControllerServer-DeleteSnapshot... ", zap.Reflect("Request", *req))
+
+	// Validate arguments
+	snapshotID := req.GetSnapshotId()
+	if len(snapshotID) == 0 {
+		return nil, commonError.GetCSIError(ctxLogger, commonError.EmptySnapshotID, requestID, nil)
+	}
+
+	// get the session
+	session, err := csiCS.CSIProvider.GetProviderSession(ctx, ctxLogger)
+	if err != nil {
+		if userError.GetUserErrorCode(err) == string(utilReasonCode.EndpointNotReachable) {
+			return nil, commonError.GetCSIError(ctxLogger, commonError.EndpointNotReachable, requestID, err)
+		}
+		if userError.GetUserErrorCode(err) == string(utilReasonCode.Timeout) {
+			return nil, commonError.GetCSIError(ctxLogger, commonError.Timeout, requestID, err)
+		}
+		return nil, commonError.GetCSIError(ctxLogger, commonError.InternalError, requestID, err)
+	}
 
-	ctxLogger.Info("CSIControllerServer-DeleteSnapshot", zap.Reflect("Request", *req))
-	return nil, commonError.GetCSIError(ctxLogger, commonError.MethodUnimplemented, requestID, nil, "DeleteSnapshot")
+	_, err = session.GetSnapshot(snapshotID)
+	if providerError.RetrivalFailed == providerError.GetErrorType(err) {
+		ctxLogger.Info("Snapshot not found. Returning success without deletion...")
+		return &csi.DeleteSnapshotResponse{}, nil
+	}
+	snapshot := &provider.Snapshot{}
+	snapshot.SnapshotID = snapshotID
+	err = session.DeleteSnapshot(snapshot)
+	if err != nil {
+		return nil, commonError.GetCSIError(ctxLogger, commonError.InternalError, requestID, err)
+	}
+	return &csi.DeleteSnapshotResponse{}, nil
 }
 
 // ListSnapshots ...
 func (csiCS *CSIControllerServer) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsRequest) (*csi.ListSnapshotsResponse, error) {
 	ctxLogger, requestID := utils.GetContextLogger(ctx, false)
 	// populate requestID in the context
-	_ = context.WithValue(ctx, provider.RequestID, requestID)
+	ctx = context.WithValue(ctx, provider.RequestID, requestID)
+	ctxLogger.Info("CSIControllerServer-ListSnapshots...", zap.Reflect("Request", *req))
+	defer metrics.UpdateDurationFromStart(ctxLogger, metrics.FunctionLabel("ListSnapshots"), time.Now())
+
```

Comment:
Is this really used by any operation today ?

Reply:
Yes we need this time calculation to check from the logs  that how much time did an operation take like create volume, create snapshot etc.

Reply:
Sorry I mean to say ListSnapshot operation ? In case of ListVolumes we concluded that is not really used by any existing external operation until we enable VolumeHealthCapabilities

Reply:
List snapshots is called by external snapshotter to update the  the status "ReadyToUse" to true on the VolumeSnapshotContent object to indicate the snapshot is ready to be used to restore a volume.